pkg/config/signer.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	publicKey, err := ecdsaPublicKeyImport(cert)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ECDSA public key: %v", err)
+	}
+
+	pkBlock, _ := pem.Decode(privateCert)
+	if pkBlock == nil {
+		return nil, errors.New("failed to decode private key from pem")
+	}
+
+	privateKey, err := ecdsaPrivateKeyImport(pkBlock.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ECDSA private key: %v", err)
+	}
+
+	signer := &Signer{
+		cert:       cert,
+		mspID:      mspID,
+		privateKey: privateKey,
+		publicKey:  publicKey,
+	}
+	return signer, nil
+}
+
+// Serialize returns a byte array representation of this identity.
+func (s *Signer) Serialize() ([]byte, error) {
+	pb := &pem.Block{Bytes: s.cert.Raw, Type: "CERTIFICATE"}
+	pemBytes := pem.EncodeToMemory(pb)
+	if pemBytes == nil {
+		return nil, errors.New("failed to encode pem block")
+	}
+
+	// serialize identities by prepending the MSPID and appending
+	// the ASN.1 DER content of the cert
+	sID := &msp.SerializedIdentity{Mspid: s.mspID, IdBytes: pemBytes}
+	idBytes := MarshalOrPanic(sID)
+
+	return idBytes, nil
+}
+
+// Public returns the public key of the signer.
+func (s *Signer) Public() crypto.PublicKey {
+	return s.publicKey
+}
+
+// Cert returns the certificate of the signer.
+func (s *Signer) Cert() *x509.Certificate {
+	return s.cert
+}
+
+// MSPId returns the MSP ID of the signer.
+func (s *Signer) MSPId() string {
+	return s.mspID
+}
+
+// Sign performs ECDSA sign with signer's private key on given digest.
+func (s *Signer) Sign(reader io.Reader, digest []byte) (signature []byte, err error) {
+	if reader == nil {
+		return nil, errors.New("failed to sign, reader can not be nil")
+	}
+
+	rr, ss, err := ecdsa.Sign(reader, s.privateKey, digest)
+	if err != nil {
+		return nil, err
+	}
+
+	ss, _, err = toLowS(&s.privateKey.PublicKey, ss)
+	if err != nil {
+		return nil, err
+	}
+
+	return asn1.Marshal(ecdsaSignature{rr, ss})
+}
+
+// CreateSignatureHeader returns a new SignatureHeader. It contains a valid
+// nonce and a creator, which is a serialized representation of the signer.
+func (s *Signer) CreateSignatureHeader() (*common.SignatureHeader, error) {
+	creator, err := s.Serialize()
+	if err != nil {
+		return nil, err
+	}
+	nonce, err := createNonce()
+	if err != nil {
+		return nil, err
+	}
+
+	return &common.SignatureHeader{
+		Creator: creator,
+		Nonce:   nonce,
+	}, nil
+}
+
+// createNonce generates a nonce using the crypto/rand package.
+func createNonce() ([]byte, error) {
+	nonce, err := getRandomNonce()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate random nonce: %s", err)
+	}
+	return nonce, nil
+}
+
+func getRandomNonce() ([]byte, error) {
+	key := make([]byte, 24)
+
+	_, err := rand.Read(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get random bytes: %s", err)
+	}
+	return key, nil
+}
+
+// isLowS checks that s is a low-S.
+func isLowS(k *ecdsa.PublicKey, s *big.Int) (bool, error) {
```

Comment:
***ecdsa.PublicKey** embeds **elliptic.Curve** which has **Params() *CurveParams** and it has the order of the curve base point which you can divide by 2. why do we need the **curveHalfOrders** array?

Reply:
Dereck and I won't pretend to have a full understanding of ecdsa. Feedback like this is much needed and appreciated!

Reply:
Done.